File: server/src/controllers/userController.ts
import { Request, Response } from 'express'
import { User } from '../models/User.js'

// ✅ Fetch all users (requires authentication)
export const getAllUsers = async (req: Request, res: Response): Promise<void> => {
  try {
    const users = await User.find().select('-password')
    res.json({ success: true, data: users })
  } catch (error: any) {
    res.status(500).json({
      success: false,
      message: error.message || 'Failed to fetch users',
    })
  }
}

// ✅ Fetch a single user by ID
export const getUserById = async (req: Request, res: Response): Promise<void> => {
  try {
    const user = await User.findById(req.params.userId).select('-password')
    if (!user) {
      res.status(404).json({
        success: false,
        message: 'User not found',
      })
      return
    }
    res.json({ success: true, data: user })
  } catch (error: any) {
    res.status(500).json({
      success: false,
      message: error.message || 'Failed to fetch user',
    })
  }
}

File: server-spring/src/main/java/com/codeeditor/server/controller/UserController.java
package com.codeeditor.server.controller;

import com.codeeditor.server.dto.ApiResponse;
import com.codeeditor.server.model.User;
import com.codeeditor.server.repository.UserRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/users")
public class UserController {

    private final UserRepository userRepository;

    public UserController(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    /** GET /api/users - Fetch all users */
    @GetMapping
    public ResponseEntity<ApiResponse> getAllUsers() {
        try {
            List<User> users = userRepository.findAll();
            // Remove password from response
            users.forEach(u -> u.setPassword(null));
            return ResponseEntity.ok(ApiResponse.success(users));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error(e.getMessage()));
        }
    }

    /** GET /api/users/:userId - Fetch a single user by ID */
    @GetMapping("/{userId}")
    public ResponseEntity<ApiResponse> getUserById(@PathVariable String userId) {
        try {
            var opt = userRepository.findById(userId);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("User not found"));
            User user = opt.get();
            user.setPassword(null);
            return ResponseEntity.ok(ApiResponse.success(user));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error(e.getMessage()));
        }
    }
}
